File: frontend/src/core/replay/ReplayRunner.ts
import type { StatePatchOperation } from '../actions/types';
import type { CompatibleGameState } from '../state/gameState';
import { createStateHash } from '../debug/stateDiff';
import type { ReplayInput } from './ReplayInput';
import type { ReplayResult } from './ReplayResult';

function cloneState(state: CompatibleGameState): CompatibleGameState {
  return JSON.parse(JSON.stringify(state || {}));
}

function setPath(target: Record<string, unknown>, path: string, value: unknown) {
  const parts = path.split('.').filter(Boolean);
  let cursor: Record<string, unknown> = target;
  for (const part of parts.slice(0, -1)) {
    if (typeof cursor[part] !== 'object' || cursor[part] === null) cursor[part] = {};
    cursor = cursor[part] as Record<string, unknown>;
  }
  cursor[parts[parts.length - 1]] = value;
}

function getPath(target: Record<string, unknown>, path: string) {
  return path.split('.').filter(Boolean).reduce<unknown>((cursor, part) => (
    typeof cursor === 'object' && cursor !== null ? (cursor as Record<string, unknown>)[part] : undefined
  ), target);
}

function removePath(target: Record<string, unknown>, path: string) {
  const parts = path.split('.').filter(Boolean);
  const key = parts.pop();
  const parent = parts.reduce<unknown>((cursor, part) => (
    typeof cursor === 'object' && cursor !== null ? (cursor as Record<string, unknown>)[part] : undefined
  ), target);
  if (key && typeof parent === 'object' && parent !== null) delete (parent as Record<string, unknown>)[key];
}

function applyPatchOperation(state: CompatibleGameState, patch: StatePatchOperation) {
  if (patch.op === 'set') setPath(state, patch.path, patch.value);
  if (patch.op === 'increment') {
    const current = Number(getPath(state, patch.path) || 0);
    setPath(state, patch.path, current + patch.value);
  }
  if (patch.op === 'merge') {
    const current = getPath(state, patch.path);
    setPath(state, patch.path, { ...(typeof current === 'object' && current ? current : {}), ...patch.value });
  }
  if (patch.op === 'append') {
    const current = getPath(state, patch.path);
    setPath(state, patch.path, [...(Array.isArray(current) ? current : []), patch.value]);
  }
  if (patch.op === 'remove') removePath(state, patch.path);
}

export function runReplay(input: ReplayInput): ReplayResult {
  const finalState = cloneState(input.initialGameState);
  const warnings: string[] = [];

  if (input.diceEvents.length === 0) {
    warnings.push('Replay input has no dice events; deterministic dice verification was skipped.');
  }

  input.statePatches.forEach((envelope) => {
    envelope.patches.forEach((patch) => applyPatchOperation(finalState, patch));
  });

  const actualStateHash = createStateHash(finalState);
  const expectedStateHash = input.expectedStateHash;
  const ok = !expectedStateHash || expectedStateHash === actualStateHash;

  return {
    ok,
    finalState,
    expectedStateHash,
    actualStateHash,
    warnings,
    mismatch: ok ? undefined : {
      index: input.statePatches.length - 1,
      actionId: input.actions[input.actions.length - 1]?.id,
      expected: expectedStateHash,
      actual: actualStateHash,
      reason: 'Final state hash does not match expectedStateHash.',
    },
  };
}

File: backend/core/replay/replay_runner.py
from __future__ import annotations

from copy import deepcopy
from typing import Any, Literal

from pydantic import BaseModel, Field

from core.debug.state_diff import create_state_hash


class ReplayInput(BaseModel):
    schemaVersion: Literal[1] = 1
    initialGameState: dict[str, Any]
    actions: list[dict[str, Any]] = Field(default_factory=list)
    diceEvents: list[dict[str, Any]] = Field(default_factory=list)
    statePatches: list[dict[str, Any]] = Field(default_factory=list)
    eventLog: list[dict[str, Any]] = Field(default_factory=list)
    rngSeed: str | None = None
    expectedStateHash: str | None = None


class ReplayResult(BaseModel):
    ok: bool
    finalState: dict[str, Any]
    expectedStateHash: str | None = None
    actualStateHash: str
    mismatch: dict[str, Any] | None = None
    warnings: list[str] = Field(default_factory=list)


def _get_path(state: dict[str, Any], path: str) -> Any:
    cursor: Any = state
    for part in [part for part in path.split(".") if part]:
        if not isinstance(cursor, dict):
            return None
        cursor = cursor.get(part)
    return cursor


def _set_path(state: dict[str, Any], path: str, value: Any) -> None:
    parts = [part for part in path.split(".") if part]
    cursor = state
    for part in parts[:-1]:
        cursor = cursor.setdefault(part, {})
    cursor[parts[-1]] = value


def _apply_patch(state: dict[str, Any], patch: dict[str, Any]) -> None:
    op = patch.get("op")
    path = str(patch.get("path", ""))
    value = patch.get("value")
    if op == "set":
        _set_path(state, path, value)
    elif op == "increment":
        _set_path(state, path, (_get_path(state, path) or 0) + value)
    elif op == "merge":
        current = _get_path(state, path)
        _set_path(state, path, {**(current if isinstance(current, dict) else {}), **value})
    elif op == "append":
        current = _get_path(state, path)
        _set_path(state, path, [*(current if isinstance(current, list) else []), value])
    elif op == "remove":
        parts = [part for part in path.split(".") if part]
        parent = _get_path(state, ".".join(parts[:-1])) if len(parts) > 1 else state
        if isinstance(parent, dict) and parts:
            parent.pop(parts[-1], None)


def run_replay(input_data: ReplayInput) -> ReplayResult:
    state = deepcopy(input_data.initialGameState)
    warnings: list[str] = []
    if not input_data.diceEvents:
        warnings.append("Replay input has no dice events; deterministic dice verification was skipped.")
    for envelope in input_data.statePatches:
        for patch in envelope.get("patches", []):
            _apply_patch(state, patch)
    actual_hash = create_state_hash(state)
    ok = input_data.expectedStateHash is None or input_data.expectedStateHash == actual_hash
    return ReplayResult(
        ok=ok,
        finalState=state,
        expectedStateHash=input_data.expectedStateHash,
        actualStateHash=actual_hash,
        mismatch=None if ok else {"reason": "Final state hash does not match expectedStateHash."},
        warnings=warnings,
    )
